src/myrtlespeech/model/hard_lstm.py

```diff
@@ -0,0 +1,419 @@
+import math
+from typing import List
+from typing import Optional
+from typing import Tuple
+
+import torch
+from myrtlespeech.model.rnn import init_hidden_state
+from myrtlespeech.model.rnn import RNNType
+from torch import Tensor
+
+
+class HardLSTM(torch.nn.Module):
```

Comment:
Proposal: We change `RNN` to use horrible types (that contain only the allowed TorchScript types, Tensor, Tuple, Union, etc) rather than `RNNState`, `RNNData`, and `Lengths` and change HardLSTM to be part of the `RNN` class (`RNNType.HARD_LSTM`). 

----

In general I'm against making the types unnecessarily complex, especially when the RNN type becomes quite long. However, TorchScript has openly stated that it has _no_ plans to support these features in the future:

> TorchScript does not support all features and types of the typing module. Some of these are more fundamental things that are unlikely to be added in the future while others may be added if there is enough user demand to make it a priority.

> Type aliases | Not implemented
> Nominal vs structural subtyping | Nominal typing is in development, but structural typing is not
> NewType | Unlikely to be implemented
> Generics | Unlikely to be implemented

> Any other functionality from the typing module not explitily listed in this documentation is unsupported.

Given this, if we want to continue using a generic RNN class to give the user a single API for all of our RNN variants whilst also having TorchScript/ONNX support then we must unfortunately deal with the complex type annotations. 

Complex type annotations can sit alongside detailed comments and docstrings. This is less of a burden than maintaining separate classes for each RNN type we support. 

Thoughts?

Reply:
Discussed in person. Leaving for now (although I agree this would be better long term).